### ISSUE TITLE
[JENKINS-2181] Multiple parallel runs per job

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/remoting#898]
<p>Currently it is possible to 'Tie this project to a node' or to a label.<br/>
A set of slaves can have the same label (so I expect it to spread the builds<br/>
over the nodes under that label if one of the nodes is busy).</p>

<p>But it is not possible to run multiple parallel runs per job on multiple nodes.</p>

<p>Use case:<br/>
We have a testrun of 2 hour but that is a pretty long feedback loop. We do more<br/>
commits. A first compile is done is a separate job to do a quick test, but the<br/>
complete run takes nearly 2 hours. The quick job runs the testrun after it when<br/>
successful. So when the testrun is finished it starts directly again with the<br/>
new svn update. </p>

<p>Disadvantage of this:</p>
<ul class="alternate" type="square">
	<li>Multiple check-ins are picked up at one</li>
	<li>It can take 4 hours before you know the results of the tests of your change</li>
	<li>Testmachine is continuously busy</li>
</ul>

<p>If the testrun can run in parallel on multiple node we can have feedback loop<br/>
per changeset in stead of a combined testset and we can take nodes offline for<br/>
maintanance.</p>

<p>So requirement:<br/>
A checkbox 'Enable parallel runs' per project (not every project can run in<br/>
parallel). Then the project needs to be tied to a set of nodes via a label or<br/>
not tie the project at all. <br/>
When all possible nodes are busy it still gets queued.</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cbos">cbos</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-2181" target="_blank">Multiple parallel runs per job</a></i></summary>
<i><ul>
<li><b>status</b>: Closed
<li><b>priority</b>: Major
<li><b>component(s)</b>: remoting
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2009-11-30T00:11:41+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 0
<li><b>imported</b>: 2025-11-25
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>Currently it is possible to 'Tie this project to a node' or to a label.
A set of slaves can have the same label (so I expect it to spread the builds
over the nodes under that label if one of the nodes is busy).</p>

<p>But it is not possible to run multiple parallel runs per job on multiple nodes.</p>

<p>Use case:
We have a testrun of 2 hour but that is a pretty long feedback loop. We do more
commits. A first compile is done is a separate job to do a quick test, but the
complete run takes nearly 2 hours. The quick job runs the testrun after it when
successful. So when the testrun is finished it starts directly again with the
new svn update. </p>

<p>Disadvantage of this:</p>
<ul class="alternate" type="square">
	<li>Multiple check-ins are picked up at one</li>
	<li>It can take 4 hours before you know the results of the tests of your change</li>
	<li>Testmachine is continuously busy</li>
</ul>

<p>If the testrun can run in parallel on multiple node we can have feedback loop
per changeset in stead of a combined testset and we can take nodes offline for
maintanance.</p>

<p>So requirement:
A checkbox 'Enable parallel runs' per project (not every project can run in
parallel). Then the project needs to be tied to a set of nodes via a label or
not tie the project at all. 
When all possible nodes are busy it still gets queued.</p></pre>
</details>
</details>
<ul><li><i>environment</i>: <code>Platform: All, OS: All</code></li></ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>gouzi:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-2181?focusedId=126108&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-126108">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I'm having the same kind of problem: we have a job1 and a downstream job2. For
each commit in our SCM we trigger job1. And for each successful build of job1,
job2 is triggered since it's a downstream project.
The thing is: job2 is taking ~2h to complete (actually in job2 we are running
the software to perform a bunch of tests). During these 2 hours many builds of
job1 are triggered and we would like to perform job2 for each build of job1. To
do so it would be nice to have an option to be able to run several parallel
builds of job2 on different available nodes.</p>

<p>Today:
<span class="error">&#91;job1-01---&#93;</span>-&gt;<span class="error">&#91;job2-01------------&#93;</span>
             <span class="error">&#91;job1-02---&#93;</span>+++++++++&gt;<span class="error">&#91;job2-02------------&#93;</span></p>

<p>It would be nice to have:
<span class="error">&#91;job1-01---&#93;</span>-&gt;<span class="error">&#91;job2-01------------&#93;</span>
             <span class="error">&#91;job1-02---&#93;</span>-&gt;<span class="error">&#91;job2-02------------&#93;</span>
Where:
'[' : job starts
']' : job is complete
'&gt;': trigger downstream project
'+': wait for the previous build of job2 to be complete</p>

<p>Since job1 can provide its whole workspace to job2, we could tell job2 to create
a new directory in its workspace based on its build number. This way we may be
able to run multiple builds in parallel for job2.</p>
</pre>
</details>

</ul>
</details>

<p>I'm having the same kind of problem: we have a job1 and a downstream job2. For<br/>
each commit in our SCM we trigger job1. And for each successful build of job1,<br/>
job2 is triggered since it's a downstream project.<br/>
The thing is: job2 is taking ~2h to complete (actually in job2 we are running<br/>
the software to perform a bunch of tests). During these 2 hours many builds of<br/>
job1 are triggered and we would like to perform job2 for each build of job1. To<br/>
do so it would be nice to have an option to be able to run several parallel<br/>
builds of job2 on different available nodes.</p>

<p>Today:<br/>
<span class="error">&#91;job1-01---&#93;</span>-><span class="error">&#91;job2-01------------&#93;</span><br/>
     <span class="error">&#91;job1-02---&#93;</span>+++++++++><span class="error">&#91;job2-02------------&#93;</span></p>

<p>It would be nice to have:<br/>
<span class="error">&#91;job1-01---&#93;</span>-><span class="error">&#91;job2-01------------&#93;</span><br/>
     <span class="error">&#91;job1-02---&#93;</span>-><span class="error">&#91;job2-02------------&#93;</span><br/>
Where:<br/>
'[' : job starts<br/>
']' : job is complete<br/>
'>': trigger downstream project<br/>
'+': wait for the previous build of job2 to be complete</p>

<p>Since job1 can provide its whole workspace to job2, we could tell job2 to create<br/>
a new directory in its workspace based on its build number. This way we may be<br/>
able to run multiple builds in parallel for job2.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>cbos:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-2181?focusedId=133723&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-133723">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>It is fixed, it is possible now to run in parallel</p>
</pre>
</details>

</ul>
</details>

<p>It is fixed, it is possible now to run in parallel</p>
